### PR TITLE
Publish raw events

### PR DESCRIPTION
**Pull request by tnsetting**
# TL;DR
Send raw flyte events to a pubsub topic. This is the first step for flyte lineage events. These events will be consumed and enriched to external by another thread or process. If the `externalEvent` part is not configured, a noop publisher will be created.

Example of config:
```
notifications:
  type: gcp
  region: "my-region"
  gcp:
    projectId: "test"
  publisher:
    topicName: "topic"
  processor:
    queueName: "queue"
    accountId: "bar"
  externalEvent:
    eventPublisher:
      topicName: "event_topic"
      eventTypes: "all" // Or workflow, node, task. Or "*"
```

## Type
 - [ ] Bug Fix
 - [x] Feature
 - [ ] Plugin

## Are all requirements met?

 - [ ] Code completed
 - [ ] Smoke tested
 - [ ] Unit tests added
 - [ ] Code documentation added
 - [ ] Any pending items have an associated Issue

## Complete description
 _How did you fix the bug, make the feature etc. Link to any design docs etc_

## Tracking Issue
https://github.com/lyft/flyte/issues/<number>

## Follow-up issue
_NA_
OR
_https://github.com/lyft/flyte/issues/<number>_

### CODE REVIEW

**Comment by tnsetting** [comment]
@kumare3 or @katrogan Any comments for the PR? Would it make more sense to have separate config section for the event? Thanks.

File: pkg/async/notifications/factory.go [hunk 1]
```diff
@@ -148,3 +148,57 @@ func NewNotificationsPublisher(config runtimeInterfaces.NotificationsConfig, sco
 		return implementations.NewNoopPublish()
 	}
 }
+
+func NewEventsPublisher(config runtimeInterfaces.NotificationsConfig, scope promutils.Scope) interfaces.Publisher {
+	if config.ExternalEvent.EventPublisherConfig.TopicName == "" {
```
Comment (katrogan) [inline pkg/async/notifications/factory.go:153]: can we explicitly just use the `Type` of the Notificationsconfig. and if it's not no-op, error when Topic Name is empty?

File: pkg/async/notifications/implementations/event_publisher.go [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+}
+
+// TODO: Add a counter that encompasses the publisher stats grouped by project and domain.
+type EventPublisher struct {
+	pub           pubsub.Publisher
+	systemMetrics eventPublisherSystemMetrics
+	events        []string
+}
+
+func getSupportedEvents() map[string]string {
```
Comment (katrogan) [inline pkg/async/notifications/implementations/event_publisher.go:31]: nit: instead of using a func you can actually initialize the map in golang as a var. see https://stackoverflow.com/questions/41078272/initializing-a-go-map-in-a-single-statement for an example

File: pkg/async/notifications/implementations/event_publisher.go [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+		for _, e := range supportedEvents {
+			eventList = append(eventList, e)
+		}
+	} else {
+		events := strings.Split(eventTypes, ",")
+		for _, event := range events {
+			if e, found := supportedEvents[event]; found {
+				eventList = append(eventList, e)
+			}
+		}
```
Comment (katrogan) [inline pkg/async/notifications/implementations/event_publisher.go:90]: can we raise an error or log if a user passes in an unsupported event?

File: pkg/async/notifications/implementations/event_publisher.go [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+		Scope:        scope,
+		PublishTotal: scope.MustNewCounter("event_publish_total", "overall count of publish messages"),
+		PublishError: scope.MustNewCounter("event_publish_errors", "count of publish errors"),
+	}
+}
+
+func NewEventsPublisher(pub pubsub.Publisher, scope promutils.Scope, eventTypes string) interfaces.Publisher {
+	supportedEvents := getSupportedEvents()
+
+	var eventList = make([]string, 0)
```
Comment (katrogan) [inline pkg/async/notifications/implementations/event_publisher.go:79]: do you want to use a[ String set](https://pkg.go.dev/k8s.io/apimachinery/pkg/util/sets#String) here instead since you only ever check for membership/

File: pkg/async/notifications/implementations/event_publisher.go [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+	supportedEvents["node"] = proto.MessageName(&nodeExecutionReq)
+	var workflowExecutionReq admin.WorkflowExecutionEventRequest
+	supportedEvents["workflow"] = proto.MessageName(&workflowExecutionReq)
+	return supportedEvents
+}
+
+// The key is the notification type as defined as an enum.
+func (p *EventPublisher) Publish(ctx context.Context, notificationType string, msg proto.Message) error {
+	p.systemMetrics.PublishTotal.Inc()
+	logger.Debugf(ctx, "Publishing the following message [%s]", msg.String())
```
Comment (katrogan) [inline pkg/async/notifications/implementations/event_publisher.go:45]: does String() return a human readable message? if not can we use `...[%+v]", msg)` instead?

File: pkg/async/notifications/implementations/event_publisher.go [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+			return true
+		}
+	}
+	return false
+}
+
+func newEventPublisherSystemMetrics(scope promutils.Scope) eventPublisherSystemMetrics {
+	return eventPublisherSystemMetrics{
+		Scope:        scope,
+		PublishTotal: scope.MustNewCounter("event_publish_total", "overall count of publish messages"),
```
Comment (katrogan) [inline pkg/async/notifications/implementations/event_publisher.go:71]: i know it's all additive, but can we add a success counter too?

File: pkg/runtime/interfaces/application_configuration.go [hunk 7]
```diff
@@ -146,6 +146,18 @@ type NotificationsEmailerConfig struct {
 	Body string `json:"body"`
 }
 
+// This section handles configuration for the workflow notifications pipeline.
+type EventsPublisherConfig struct {
+	// The topic which events should be published, e.g. node, task, workflow
+	TopicName string `json:"topicName"`
+	// Event types: task, node, workflow executions
+	EventTypes string `json:"eventTypes"`
```
Comment (katrogan) [inline pkg/runtime/interfaces/application_configuration.go:154]: why not an array type?

**Review by katrogan — COMMENTED** [review]
looks good! just a few suggestions

**Comment by katrogan** [comment]
> Would it make more sense to have separate config section for the event instead of inside the notification section? Thanks.

Yes let's please do that! I think we can worry about refactoring the existing email notifications implementation separately from the egress event work you're doing. Eventually notifications can be handled only by a separately configured subscriber so it make sense to separate out the events config now!

File: pkg/runtime/interfaces/application_configuration.go (hunk 7 above)
Comment (tnsetting) [reply pkg/runtime/interfaces/application_configuration.go:154]: Because I want to support the option to use `all` or `*` and using a single string makes it easier to filter on such cases.

File: pkg/async/notifications/implementations/event_publisher.go (hunk 3 above)
Comment (tnsetting) [reply pkg/async/notifications/implementations/event_publisher.go:91]: Error loggings added.

**Comment by katrogan** [comment]
hey @tnsetting is this ready for re-review?

**Comment by tnsetting** [comment]
> hey @tnsetting is this ready for re-review?

Yes, it is ready.

File: pkg/runtime/interfaces/application_configuration.go (hunk 7 above)
Comment (katrogan) [reply pkg/runtime/interfaces/application_configuration.go:154]: but you can still do that with a single entry, no? I'd prefer to use conventional yaml list formatting vs our own string-split scheme

File: pkg/async/notifications/implementations/event_publisher.go [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+	systemMetrics eventPublisherSystemMetrics
+	events        sets.String
+}
+
+var taskExecutionReq admin.TaskExecutionEventRequest
+var nodeExecutionReq admin.NodeExecutionEventRequest
+var workflowExecutionReq admin.WorkflowExecutionEventRequest
+
+var supportedEvents = map[string]string{
+	"task":     proto.MessageName(&taskExecutionReq),
```
Comment (katrogan) [inline pkg/async/notifications/implementations/event_publisher.go:39]: sorry should have brought this up in the last round, but can we use shared constants here as map keys? perhaps https://github.com/flyteorg/flyteadmin/blob/master/pkg/common/entity.go could work

File: pkg/async/notifications/implementations/event_publisher.go [hunk 9]
```diff
@@ -0,0 +1,98 @@
...
+	"node":     proto.MessageName(&nodeExecutionReq),
+	"workflow": proto.MessageName(&workflowExecutionReq),
+}
+
+// The key is the notification type as defined as an enum.
+func (p *EventPublisher) Publish(ctx context.Context, notificationType string, msg proto.Message) error {
+	p.systemMetrics.PublishTotal.Inc()
+	logger.Debugf(ctx, "Publishing the following message [%+v]", msg)
```
Comment (katrogan) [inline pkg/async/notifications/implementations/event_publisher.go:47]: nit: do you want to move this log statement below the check to shouldPublishEvent?

File: pkg/async/notifications/implementations/event_publisher_test.go [hunk 10]
```diff
@@ -0,0 +1,173 @@
...
+}
+
+// This method should be invoked before every test around Publisher.
+func initializeEventPublisher() {
+	testEventPublisher.Published = nil
+	testEventPublisher.GivenError = nil
+	testEventPublisher.FoundError = nil
+}
+
+func TestNewEventsPublisher_EventTypes(t *testing.T) {
```
Comment (katrogan) [inline pkg/async/notifications/implementations/event_publisher_test.go:90]: Thank you for this very exhaustive test 👍

File: pkg/manager/impl/node_execution_manager.go [hunk 11]
```diff
@@ -4,6 +4,9 @@ import (
 	"context"
 	"strconv"
 
+	"github.com/gogo/protobuf/proto"
```
Comment (katrogan) [inline pkg/manager/impl/node_execution_manager.go:7]: can we use `github.com/golang/protobuf/proto` instead? gogo behaves weirdly with oneofs

**Review by katrogan — COMMENTED** [review]
sorry just a few more comments

File: pkg/manager/impl/node_execution_manager.go (hunk 11 above)
Comment (tnsetting) [reply pkg/manager/impl/node_execution_manager.go]: Sure. Never paid attention for that as intellij did that for me.

File: pkg/async/notifications/implementations/event_publisher.go [hunk 12]
```diff
@@ -0,0 +1,105 @@
...
+	systemMetrics eventPublisherSystemMetrics
+	events        sets.String
+}
+
+var taskExecutionReq admin.TaskExecutionEventRequest
+var nodeExecutionReq admin.NodeExecutionEventRequest
+var workflowExecutionReq admin.WorkflowExecutionEventRequest
+
+const (
+	Task          = "task"
```
Comment (katrogan) [inline pkg/async/notifications/implementations/event_publisher.go:38]: I'm still concerned this is a little brittle because here we define the strings statically but in the manager methods we use the proto message name as the event type. does that correspond to these raw strings like "task" or "node"?

**Review by katrogan — COMMENTED** [review]
do you mind updating https://github.com/flyteorg/flyteadmin/blob/master/flyteadmin_config.yaml with an example of the config changes too?

File: pkg/async/notifications/implementations/event_publisher.go (hunk 12 above)
Comment (katrogan) [reply pkg/async/notifications/implementations/event_publisher.go:38]: @tnsetting

**Review by katrogan — APPROVED** [review]
THANK YOU so much for this change
